src/h3c_fmt_plug.c

```diff
@@ -0,0 +1,320 @@
+	base64_convert(ciphertext, e_b64_mime, strlen(ciphertext), realcipher, e_b64_raw, sizeof(x.out), flg_Base64_NO_FLAGS, 0);
+
+#if defined(SIMD_COEF_64) && ARCH_LITTLE_ENDIAN==1
+	alter_endianity_to_BE64(realcipher, DIGEST_SIZE / 8);
+#endif
+	return (void *)realcipher;
+}
+
+static void set_salt(void *salt)
+{
+	memcpy(cursalt, salt, SALT_SIZE);
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static char salt[SALT_SIZE];
+
+	memcpy(salt, ciphertext + TAG_LENGTH, SALT_SIZE);
+	return (void *)salt;
+}
+
+#define COMMON_GET_HASH_64BIT_HASH
+#define COMMON_GET_HASH_VAR crypt_out
+#include "common-get-hash.h"
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	int count = *pcount;
+	int index;
+
+#ifdef _OPENMP
+	#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += NBKEYS) {
+#ifdef SIMD_COEF_64
+		unsigned char _in[8 * 16 * MIN_KEYS_PER_CRYPT + MEM_ALIGN_SIMD];
+		unsigned char _out[8 * 8 * MIN_KEYS_PER_CRYPT + MEM_ALIGN_SIMD];
+		uint64_t *in = (uint64_t *) mem_align(_in, MEM_ALIGN_SIMD);
+		uint64_t *out = (uint64_t *) mem_align(_out, MEM_ALIGN_SIMD);
+
+		for (int i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
+			int x80_off = (saved_len[index + i] << 1) + 2 + SALT_SIZE;
+			unsigned char *cp = (unsigned char *) & (in[16 * i]);
+			int key_length = saved_len[index + i] + 1;      //include null byte
+
+			memcpy(cp, saved_key[index + i], key_length);
+			memcpy(&cp[key_length], cursalt, SALT_SIZE);
+			memcpy(&cp[key_length + SALT_SIZE], saved_key[index + i], key_length);
+			cp[x80_off] = 0x80;
+			memset(&cp[x80_off + 1], 0, 120 - (x80_off + 1));
+			in[i * 16 + 15] = x80_off << 3;
+		}
+
+		SIMDSHA512body(in, out, NULL, SSEi_FLAT_IN | SSEi_FLAT_OUT);
+
+		for (int i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
+			for (int j = 0; j < DIGEST_SIZE / sizeof(uint64_t); ++j) {
+#if ARCH_LITTLE_ENDIAN==1
+				crypt_out[index + i][j] = JOHNSWAP64(out[i * (DIGEST_SIZE / sizeof(uint64_t)) + j]);
+#else
+				crypt_out[index + i][j] = out[i * (DIGEST_SIZE / sizeof(uint64_t)) + j];
+#endif
+			}
+		}
+#else
+		SHA512_CTX ctx;
+
+		SHA512_Init(&ctx);
+		SHA512_Update(&ctx, saved_key[index], saved_len[index] + 1);    // include null byte
+		SHA512_Update(&ctx, cursalt, SALT_SIZE);
+		SHA512_Update(&ctx, saved_key[index], saved_len[index] + 1);    // include null byte
+		SHA512_Final((unsigned char *)crypt_out[index], &ctx);
+#endif
+	}
+
+	return count;
+}
+
+static void set_key(char *key, int index)
+{
+	saved_len[index] = strnzcpyn(saved_key[index], key, sizeof(saved_key[index]));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int cmp_all(void *binary, int count)
+{
+	int index;
+
+	for (index = 0; index < count; index++)
+		if (((uint64_t *) binary)[0] == crypt_out[index][0])
+			return 1;
+	return 0;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return *(uint64_t *) binary == crypt_out[index][0];
+}
+
+static int cmp_exact(char *source, int index)
+{
+	uint64_t *binary = get_binary(source);
```

Comment:
I think we don't need a fully implemented `cmp_exact` at all. We could instead compare the full computed hash in `cmp_one`.

Reply:
We need this as-is as long as the format use a reduced binary size (8 of 64). That's not needed for this format though, so the fix could be to stop using that, which would also get rid of some copying in SIMD crypt_all. Then cmp_one would be just a full memcmp and cmp_exact would just return 1.